File: day11/day11.js
import { rawInput } from './input.js';
// import {exampleInput as rawInput} from './input.js';
import IntCode from './IntCode.js';

const program = rawInput.split(',').map(Number);

function printCode(pixels) {
    const canvas = document.getElementById('canvas');
    const ctx = canvas.getContext('2d');
    ctx.fillStyle = 'red';
    const size = 6;

    pixels.forEach((line, yIndex) => {
        line.forEach((pixel, xIndex) => {
            if (pixel === '#') {
                ctx.fillRect(xIndex * size, yIndex * size, size, size);
            }
        });
    });
}

function paintHull(startingInput) {
    const hull = new Map();
    const currPos = [0, 0];
    let direction = 0;
    const DIRECTIONS = { 0: [1, 0], 1: [0, 1], 2: [-1, 0], 3: [0, -1] };

    const robot = new IntCode('Robot', program, startingInput);

    let output = robot.run();
    while (output !== 99) {
        const [paint, newDir] = output;
        hull.set(currPos.toString(), paint);

        if (newDir === 0) {
            direction = direction === 0 ? 3 : direction - 1;
        } else {
            direction = direction === 3 ? 0 : direction + 1;
        }
        currPos[0] += DIRECTIONS[direction][0];
        currPos[1] += DIRECTIONS[direction][1];
        robot.processInput(hull.get(currPos.toString()) ?? 0);
        output = robot.run();
    }
    return hull;
}

const partOne = paintHull(0);
const partTwo = paintHull(1);

const grid = [];
for (let i = 0; i < 6; i++) {
    grid.push(new Array(40));
    for (let j = 0; j < 40; j++) {
        grid[i][j] = ' ';
    }
}

partTwo.forEach((paint, coords) => {
    if (paint) {
        const [x, y] = coords.split(',').map(Number);
        grid[Math.abs(x)][y] = '#';
    }
});

console.log(partOne.size);
console.log(grid);
document.getElementById('partOne').appendChild(document.createTextNode(partOne.size));
printCode(grid);

File: day11/day11.py
from collections import defaultdict
from intcode import Intcode

LEFT, RIGHT = 0, 1
NORTH, SOUTH, EAST, WEST = ("N", "S", "E", "W")

DIRMAP = {
    NORTH: (WEST, EAST),
    SOUTH: (EAST, WEST),
    EAST: (NORTH, SOUTH),
    WEST: (SOUTH, NORTH),
}

MOVEMAP = {NORTH: (-1, 0), SOUTH: (1, 0), EAST: (0, 1), WEST: (0, -1)}


def parse_input():
    with open("input.txt") as f:
        return [int(n) for n in f.read().split(",")]


def sparse_to_matrix(grid):
    min_x = min(x for x, _ in grid)
    max_x = max(x for x, _ in grid)
    min_y = min(y for _, y in grid)
    max_y = max(y for _, y in grid)

    height = max_x - min_x + 1
    width = max_y - min_y + 1

    matrix = [([" "] * width) for _ in range(height)]

    for x in range(height):
        for y in range(width):
            if grid[min_x + x, min_y + y] == 1:
                matrix[x][y] = "#"

    return matrix


def paintHull(startingInput):
    pos = (0, 0)
    dir = "N"
    hull = defaultdict(lambda: 0)

    robot = Intcode(program)
    output = robot.run(startingInput)

    while output != 99:
        color, newDir = output
        hull[pos] = color

        dir = DIRMAP[dir][newDir]
        dx, dy = MOVEMAP[dir]
        pos = (pos[0] + dx, pos[1] + dy)
        output = robot.run(hull[pos])

    return hull


program = parse_input()
partOne = paintHull(0)
print(len(partOne))
partTwo = paintHull(1)
hull = sparse_to_matrix(partTwo)
for row in hull:
    print("".join(row))
